### PR TITLE
feat: Add option to use other mappers

### PR DESCRIPTION
**Pull request by latonz**
# Add option to use other mappers

## Description

Fixes https://github.com/riok/mapperly/discussions/37

Adds `UseMapper`, `UseStaticMapper` and `UseStaticMapper<T>` attributes.
`UseMapper` can be applied to any field/property, `UseStaticMapper` to the mapper class.
Mapperly will consider all accessible mapping methods found inside the referenced mappers.

### Sample
<details>

```c#

record Fruit();
record FruitDto();
record Car();
record CarDto();
class SourceContainer { public Fruit Fruit; public Car Car; }
class TargetContainer { public FruitDto Fruit; public CarDto Car; }

[UseStaticMapper<FruitMapper>]
partial class MyMapper
{
    [UseMapper]
    private readonly CarMapper _carMapper = new();

    public partial TargetContainer Map(SourceContainer source)
    // generated implementation:
    {
        var target = new TargetContainer();
        target.Fruit = FruitMapper.ToDto(source.Fruit);
        target.Car = _carMapper.ToDto(source.Car);
        return target;
    }
}

class FruitMapper
{
    public static FruitDto ToDto(Fruit fruit) => new FruitDto();
}

class CarMapper
{
    public CarDto ToDto(Car car) => new CarDto();
}

```

</details>

## Checklist

- [x] Resolve ToDo's in the code
- [x] Handle if the field/property to which the `UseMapper` is applied is nullable
- [x] Cleanups
- [x] The existing code style is followed
- [x] The commit message follows our guidelines
- [x] Performed a self-review of my code
- [x] Hard-to-understand areas of my code are commented
- [x] The documentation is updated (as applicable)
- [x] Unit tests are added/updated
- [x] Integration tests are added/updated (as applicable, especially if feature/bug depends on roslyn or framework version in use)

### CODE REVIEW

File: docs/docs/configuration/user-implemented-methods.mdx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+// highlight-end
+public static partial class BoxMapper
+{
+  public static partial BananaBox MapBananaBox(BananaBoxDto dto);
+}
+
+public static class BananaMapper
+{
+  public static Banana MapBanana(BananaDto dto)
+  => new Banana(dto.Weigth);
```
Comment (CommonGuy) [inline docs/docs/configuration/user-implemented-methods.mdx:50]: Nitpick: Indentation

File: src/Riok.Mapperly/Configuration/AttributeDataAccessor.cs [hunk 2]
```diff
@@ -161,13 +163,20 @@ is InvocationExpressionSyntax
...
-    private static object?[] BuildArrayValue(TypedConstant arg, Type targetType)
+    private static IReadOnlyCollection<object?> BuildArrayValue(TypedConstant arg, Type targetType)
     {
         if (!targetType.IsGenericType || targetType.GetGenericTypeDefinition() != typeof(IReadOnlyCollection<>))
             throw new InvalidOperationException($"{nameof(IReadOnlyCollection<object>)} is the only supported array type");
 
         var elementTargetType = targetType.GetGenericArguments()[0];
-        return arg.Values.Select(x => BuildArgumentValue(x, elementTargetType, null)).ToArray();
+        var listType = typeof(List<>).MakeGenericType(elementTargetType);
+        var list = (IList)Activator.CreateInstance(listType, arg.Values.Length);
```
Comment (CommonGuy) [inline src/Riok.Mapperly/Configuration/AttributeDataAccessor.cs:173]: FYI: All of this does not seem to be covered by tests

File: test/Riok.Mapperly.Tests/Mapping/UseMapperTest.cs [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+            );
+    }
+
+    [Fact]
+    public void DisabledNullableShouldWork()
+    {
+        var source = TestSourceBuilder.MapperWithBodyAndTypes(
+            """
+            [UseMapper]
+            private readonly OtherMapper _otherMapper;
```
Comment (CommonGuy) [inline test/Riok.Mapperly.Tests/Mapping/UseMapperTest.cs:267]: Could you add a test where the used mapper is a property instead of a field?

Comment (latonz) [reply test/Riok.Mapperly.Tests/Mapping/UseMapperTest.cs:293]: Added, thanks for noting 😊

File: docs/docs/configuration/user-implemented-methods.mdx (hunk 1 above)
Comment (latonz) [reply docs/docs/configuration/user-implemented-methods.mdx]: Fixed.

File: src/Riok.Mapperly/Configuration/AttributeDataAccessor.cs (hunk 2 above)
Comment (latonz) [reply src/Riok.Mapperly/Configuration/AttributeDataAccessor.cs]: Reverted.

**Review by CommonGuy — DISMISSED** [review]

**Review by CommonGuy — APPROVED** [review]

**Pull request merged**